src/hotspot/cpu/x86/stubGenerator_x86_64_poly.cpp

```diff
@@ -0,0 +1,1036 @@
+  __ movq(t1, rdx);
+
+  // t3:t2 += (a1 * r0)
+  __ movq(rax, r0);
+  __ mulq(a1);
+  __ addq(t2, rax);
+  __ adcq(t3, rdx);
+
+  // t1:a0 += (a1 * r1x5)
+  __ movq(rax, c1);
+  __ mulq(a1);
+  __ addq(a0, rax);
+  __ adcq(t1, rdx);
+
+  // Note: a2 is clamped to 2-bits,
+  //       r1/r0 is clamped to 60-bits,
+  //       their product is less than 2^64.
+
+  if (only128) { // Accumulator only 128 bits, i.e. a2 == 0
+    // just move and add t1-t2 to a1
+    __ movq(a1, t1);
+    __ addq(a1, t2);
+    __ adcq(t3, 0);
+  } else {
+    // t3:t2 += (a2 * r1x5)
+    __ movq(a1, a2); // use a1 for a2
+    __ imulq(a1, c1);
+    __ addq(t2, a1);
+    __ adcq(t3, 0);
+
+    __ movq(a1, t1); // t1:a0 => a1:a0
+
+    // t3:a1 += (a2 * r0):t2
+    __ imulq(a2, r0);
+    __ addq(a1, t2);
+    __ adcq(t3, a2);
+  }
+
+  // At this point, 3 64-bit limbs are in t3:a1:a0
+  // t3 can span over more than 2 bits so final partial reduction step is needed.
+  //
+  // Partial reduction (just to fit into 130 bits)
+  //    a2 = t3 & 3
+  //    k = (t3 & ~3) + (t3 >> 2)
+  //         Y    x4  +  Y    x1
+  //    a2:a1:a0 += k
+  //
+  // Result will be in a2:a1:a0
+  __ movq(t1, t3);
+  __ movl(a2, t3); // DWORD
+  __ andq(t1, ~3);
+  __ shrq(t3, 2);
+  __ addq(t1, t3);
+  __ andl(a2, 3); // DWORD
+
+  // a2:a1:a0 += k (kept in t1)
+  __ addq(a0, t1);
+  __ adcq(a1, 0);
+  __ adcl(a2, 0); // DWORD
+}
+
+// Convert array of 128-bit numbers in quadwords (in D0:D1) into 128-bit numbers across 44-bit limbs (in L0:L1:L2)
+// Optionally pad all the numbers (i.e. add 2^128)
+//
+//         +-------------------------+-------------------------+
+//  D0:D1  | h0 h1 g0 g1 f0 f1 e0 e1 | d0 d1 c0 c1 b0 b1 a0 a1 |
+//         +-------------------------+-------------------------+
+//         +-------------------------+
+//  L2     | h2 d2 g2 c2 f2 b2 e2 a2 |
+//         +-------------------------+
+//         +-------------------------+
+//  L1     | h1 d1 g1 c1 f1 b1 e1 a1 |
+//         +-------------------------+
+//         +-------------------------+
+//  L0     | h0 d0 g0 c0 f0 b0 e0 a0 |
+//         +-------------------------+
+//
+void StubGenerator::poly1305_limbs_avx512(
+    const XMMRegister D0, const XMMRegister D1,
+    const XMMRegister L0, const XMMRegister L1, const XMMRegister L2, bool padMSG)
+{
+  const XMMRegister TMP1 = xmm0;
+  const XMMRegister TMP2 = xmm1;
+  const Register rscratch = r13;
+
+  // Interleave blocks of data
+  __ evpunpckhqdq(TMP1, D0, D1, Assembler::AVX_512bit);
+  __ evpunpcklqdq(L0, D0, D1, Assembler::AVX_512bit);
+
+  // Highest 42-bit limbs of new blocks
+  __ vpsrlq(L2, TMP1, 24, Assembler::AVX_512bit);
+  if (padMSG) {
+    __ evporq(L2, L2, ExternalAddress(poly1305_pad_msg()), Assembler::AVX_512bit, rscratch); // Add 2^128 to all 8 final qwords of the message
+  }
+
+  // Middle 44-bit limbs of new blocks
+  __ vpsrlq(L1, L0, 44, Assembler::AVX_512bit);
+  __ vpsllq(TMP2, TMP1, 20, Assembler::AVX_512bit);
```

Comment:
Thanks for the catch. Removed TMP2. (Several refactors ago, `D[01]` and `L[0-2]` used the same registers, because I was running out.. likely forgot to cleanup after I removed 2/3 of the optimizations and re-did register allocation)

done